Add how to for DateOnly and TimeOnly (#33270)
dotnet/docs: docs/standard/datetime/choosing-between-datetime.md

---
title: Compare types related to date and time

> description: Learn about differences between the DateTime, DateOnly, DateTimeOffset, TimeSpan, TimeOnly and TimeZoneInfo types to represent date and time information in .NET.

Reviewer: ```suggestion
description: Learn about differences between the DateTime, DateOnly, DateTimeOffset, TimeSpan, TimeOnly, and TimeZoneInfo types to represent date and time information in .NET.
```

> [!IMPORTANT]
> When saving or sharing <xref:System.DateTime> data, UTC should be used and the <xref:System.DateTime> value's <xref:System.DateTime.Kind%2A> property should be set to <xref:System.DateTimeKind.Utc?displayProperty=nameWithType>.

## The DateOnly structure

> The <xref:System.DateOnly> structure represents the specific date, without time. Since it has no time component, it represents a date from the start of the day to the end of the day. This structure is ideal for storing specific dates, such as a birth date, anniversary date, or business-related dates.

Reviewer: ```suggestion
The <xref:System.DateOnly> structure represents a specific date, without time. Since it has no time component, it represents a date from the start of the day to the end of the day. This structure is ideal for storing specific dates, such as a birth date, an anniversary date, or business-related dates.
```

Author: I had some copy paste errors via old copies on these. I'll update them to match the other article's content.

You may wonder why you shouldn't just use `DateTime` while ignoring the time component. There are a few benefits to using `DateOnly` over `DateTime`:

> - The `DateTime` structure may roll into the previous or next day if it's offset by a timezone. `DateOnly` can't be offset by a timezone, and it always represents the date that was set.

Reviewer: ```suggestion
- The `DateTime` structure may roll into the previous or next day if it's offset by a time zone. `DateOnly` can't be offset by a time zone, and it always represents the date that was set.
```

- Serializing a `DateTime` structure includes the time component, which may obscure the intent of the data. Also, `DateOnly` serializes less data.

> - When code interacts with a database, such as SQL Server, whole dates are generally stored as the **date** data type, which doesn't include a time. `DateOnly` is better matched to the database's type.

Reviewer: Should **date** be code-fenced for localization?

[!code-csharp[Conceptual.ChoosingDates#2](../../../samples/snippets/csharp/VS_Snippets_CLR/conceptual.choosingdates/cs/datetimereplacement1.cs#2)]
[!code-vb[Conceptual.ChoosingDates#2](../../../samples/snippets/visualbasic/VS_Snippets_CLR/conceptual.choosingdates/vb/datetimereplacement1.vb#2)]

## The TimeOnly structure

> The <xref:System.TimeOnly> structure represents a time-of-day value, such as an alarm for when to wake up a daily wake up time or what time you eat lunch each day. `TimeOnly` is limited to range of **00:00:00.0000000** to **23:59:59.9999999**, a specific time of day.

Reviewer: ```suggestion
The <xref:System.TimeOnly> structure represents a time-of-day value, such as an alarm for a daily wake up time or what time you eat lunch each day. `TimeOnly` is limited to the range of **00:00:00.0000000** to **23:59:59.9999999**, which represents a specific time of day.
```

Prior to the `TimeOnly` type being introduced, programmers typically used either the <xref:System.DateTime> structure or the <xref:System.TimeSpan> structure to represent a specific time. However, using these structures to simulate a "time only" structure has some drawbacks, which `TimeOnly` solves:

- `TimeSpan` represents elapsed time, such as time measured with a stopwatch. The upper range is more than 29,000 years, and its value can be negative to indicate moving backwards in time.

- If `TimeSpan` is used as a time of day, there's a risk that it could be manipulated to a value outside of the 24-hour day. `TimeOnly` doesn't have this risk. For example, if an employee's work shift starts at 18:00 and lasts for 8 hours, adding 8 hours to the `TimeOnly` structure rolls over to 2:00

> - Using `DateTime` for a time of day, requires that an arbitrary date be associated with the time, and later disregarded. It's common practice to choose `DateTime.MinValue` (0001-01-01) as the date, however, if hours are subtracted from the `DateTime` value, an `OutOfRange` exception may occur. `TimeOnly` doesn't have this problem as the time rolls forwards and backwards around the 24-hour time span.

Reviewer: ```suggestion
- Using `DateTime` for a time of day requires that an arbitrary date be associated with the time, and then later disregarded. It's common practice to choose `DateTime.MinValue` (0001-01-01) as the date, however, if hours are subtracted from the `DateTime` value, an `OutOfRange` exception might occur. `TimeOnly` doesn't have this problem as the time rolls forwards and backwards around the 24-hour time span.
```
